march_hardware/src/PDOmap.cpp

```diff
@@ -6,129 +6,153 @@
   }
 }
 
-std::map<enum IMCObjectName, int> PDOmap::map(int slaveIndex, enum dataDirection direction)
+std::map<IMCObjectName, int> PDOmap::map(int slave_index, enum dataDirection direction)
 {
-  this->sortPDOObjects();
-  int reg;
-  int SMAddress;
   if (direction == dataDirection::miso)
   {
-    reg = 0x1A00;
-    SMAddress = 0x1C13;
+    return configurePDO(slave_index, 0x1A00, 0x1C13);
   }
   else if (direction == dataDirection::mosi)
   {
-    reg = 0x1600;
-    SMAddress = 0x1C12;
+    return configurePDO(slave_index, 0x1600, 0x1C12);
   }
   else
   {
-    ROS_ERROR("Invalid dataDirection argument");
+    throw std::runtime_error("Invalid dataDirection argument in PDO mapping");
   }
-  // Clear SyncManager Object
-  sdo_bit8(slaveIndex, SMAddress, 0, 0);
-  int startReg = reg;
-  int lastFilledReg = reg;
-  int sizeleft = this->bitsPerReg;
-  int counter = 0;
-  int byteOffset = 0;
-  while (this->sortedPDOObjects.size() > 0)
+}
+
+std::map<IMCObjectName, int> PDOmap::configurePDO(int slave_index, int base_register, int base_sync_manager)
+{
+  int counter = 1;
+  int current_register = base_register;
+  int size_left = this->bits_per_register;
+
+  std::map<IMCObjectName, int> byte_offsets;
+  std::vector<std::pair<IMCObjectName, IMCObject>> sorted_PDO_objects = this->sortPDOObjects();
+
+  sdo_bit8(slave_index, current_register, 0, 0);
+  for (const auto& nextObject : sorted_PDO_objects)
   {
-    // Check if register is still empty
-    if (sizeleft == this->bitsPerReg)
+    size_left -= nextObject.second.length;
+    if (size_left < 0)
     {
-      sdo_bit32(slaveIndex, reg, 0, 0);
+      // PDO is filled so it can be enabled again
+      sdo_bit8(slave_index, current_register, 0, counter - 1);
+
+      // Update the sync manager with the just configured PDO
+      sdo_bit8(slave_index, base_sync_manager, 0, 0);
+      int currentPDONr = (current_register - base_register) + 1;
+      sdo_bit16(slave_index, base_sync_manager, currentPDONr, current_register);
+
+      // Move to the next PDO register by incrementing with one
+      current_register++;
+      if (current_register > (base_register + nr_of_regs))
+      {
+        ROS_ERROR("Amount of registers was overwritten, amount of parameters does not fit in the PDO messages.");
+      }
+
+      size_left = this->bits_per_register - nextObject.second.length;
+      counter = 1;
+
+      sdo_bit8(slave_index, current_register, 0, 0);
     }
-    // Get next object (from end, because sorted from small to large)
-    std::pair<IMCObjectName, IMCObject> nextObject = this->sortedPDOObjects.back();
-    this->sortedPDOObjects.pop_back();
-    // Add next object to map
+
+    int byteOffset = (bits_per_register - (size_left + nextObject.second.length)) / 8;
+    byte_offsets[nextObject.first] = byteOffset;
+
+    sdo_bit32(slave_index, current_register, counter,
+              PDOmap::combineAddressLength(nextObject.second.address, nextObject.second.length));
```

Comment:
This method only applies to the `IMCObject` class, so I think it is better to make this a member function of the class. Then you call it with `nextObject.second.combineAddressLength()`. Furthermore, the method signature uses `uint16_t` while the members are of type `int`. So change the members' type too (see other comment).